lib/sqlalchemy/dialects/sqlite/base.py

```diff
@@ -402,6 +402,253 @@ def set_sqlite_pragma(dbapi_connection, connection_record):
+to silently skip this row.
+
+Conflicts are determined using existing uniqueness constraints.
+These constraints may be identified either using their column name as stated
+in DDL, or they may be *inferred* by stating the columns and conditions that
+comprise the indexes.
+
+A "uniqueness constraint" is an explicit UNIQUE or PRIMARY KEY constraint
+within the CREATE TABLE statement, or a unique index.
+
+SQLAlchemy provides ``ON CONFLICT`` support via the SQLite-specific
+:func:`_sqlite.insert()` function, which provides
+the generative methods :meth:`~.sqlite.Insert.on_conflict_do_update`
+and :meth:`~.sqlite.Insert.on_conflict_do_nothing`::
+
+    from sqlalchemy.dialects.sqlite import insert
+
+    insert_stmt = insert(my_table).values(
+        id='some_existing_id',
+        data='inserted value')
+
+    do_nothing_stmt = insert_stmt.on_conflict_do_nothing(
+        index_elements=['id']
+    )
+
+    conn.execute(do_nothing_stmt)
+
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        constraint='id',
+        set_=dict(data='updated value')
+    )
+
+    conn.execute(do_update_stmt)
+
+Both methods supply the "target" of the conflict using either the
+named constraint or by column inference:
+
+* The :paramref:`.Insert.on_conflict_do_update.index_elements` argument
+  specifies a sequence containing string column names, :class:`_schema.Column`
+  objects, and/or SQL expression elements, which would identify a unique
+  index::
+
+
+    unique_index = schema.Index(
+        "unique_index_name",
+        my_table.c.id,
+        unique=True,
+    )
+
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        index_elements=['id'],
+        set_=dict(data='updated value')
+    )
+
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        index_elements=[my_table.c.id],
+        set_=dict(data='updated value')
+    )
+
+* When using :paramref:`.Insert.on_conflict_do_update.index_elements` to
+  infer an index, a partial index can be inferred by also specifying the
+  use the :paramref:`.Insert.on_conflict_do_update.index_where` parameter::
+
+    from sqlalchemy.dialects.sqlite import insert
+
+    unique_partial_index = schema.Index(
+        "unique_partial_index_name",
+        my_table.c.user_email,
+        unique=True,
+        sqlite_where=my_table.c.user_email.like('@gmail.com'),
+    )
+
+    stmt = insert(my_table).values(user_email='a@b.com', data='inserted data')
+    stmt = stmt.on_conflict_do_update(
+        index_elements=[my_table.c.user_email],
+        index_where=my_table.c.user_email.like('%@gmail.com'),
+        set_=dict(data=stmt.excluded.data)
+        )
+    conn.execute(stmt)
+
+* The :paramref:`.Insert.on_conflict_do_update.constraint` argument is
+  used to specify a constraint directly rather than inferring it.  This can
+  either be the name of the underlying columns of a UNIQUE constraint or
+  a PRIMARY KEY constraint. SQLite does not support specifying the names of
```

Comment:
**mike bayer** (zzzeek) wrote:

how is it different to use "constraint='id'" vs. "index_elements=['id']" ?   If these are the same thing then I think we'd just stick with index_elements for the string version and have "constraint" only point to a constraint object.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2230